keps/sig-cli/kustomize-reorder-plugins.md

```diff
@@ -0,0 +1,217 @@
+ - "@jbrette"
+creation-date: 2019-09-01
+last-updated: 2019-09-01
+status: proof of concept
+see-also:
+replaces:
+superseded-by:
+ - n/a
+---
+
+# Kustomize Resource Ordering
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Kustomize orders Resource creation by sorting the Resources it emits based off their type
+or the order used in the original files. The user needs to be able to update that order
+to account for his CRDs and the kubectl operation he intends to perform (apply, delete, patch...)
+
+See [kubernetes-sigs/kustomize#836] for an example.
+See [kubernetes-sigs/kustomize#821] for an example.
+
+## Motivation
+
+Users may need direct control of the Resource create / update / delete ordering in cases were
+sorting by Resource Type is insufficient.
+
+### Goals
+
+- Provide the ability for users to override the order that Kustomize emits Resources
+  - Used by `kubectl apply`
+  - Used by `kubectl delete`
+  
+### Non-Goals
+
+- Ensure certain Resources are *Settled* or *Ready* before other Resources
+- Ensure dependencies between Workloads
+
+## Proposal
+
+Provide a simple mechanism allowing users to override the order that
+Resource operations are applied or deleted. Leverage the plugins feature of kustomize
+in order to let the user override the default sorting plugin.  
+
+```bash
+kustomize build <xxx> --reorder=kubectlapply | kubectl apply -f -
+kustomize build <xxx> --reorder=kubectldelete | kubect delete -f -
+```
+
+or 
+
+```bash
+kubectl apply -k <xxx>
+kubectl delete -k <xxx>
+```
+
+### kubectl apply example:
+
+The `--reorder=kubectctlapply` option indicates to kustomize to load the `kubectlapplyordertransformer.yaml transformer config` ,
+itself loading a kustomize external plugin in charge of sorting and filtering kustomize build output.
+
+An example of configuration of such a kubectapplyordertransformer.yaml:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyOrderTransformer
+metadata:
+  name: kubectlapplyordertransformer
+kindorder:
+- Namespace
+- CustomResourceDefinition
+- ServiceAccount
+- ClusterRole
+- RoleBinding
+- ClusterRoleBinding
+- ConfigMap
+- Service
+- Deployment
+- CronJob
+- ValidatingWebhookConfiguration
+- APIService
+- Job
+- Certificate
+- ClusterIssuer
+- Issuer
+```
+
+A new transformer could also be implemented to filter resources during `kubectl apply` for instance:
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: KubectlApplyFilterTransformer
```

Comment:
Can you elaborate this transformer? Does it only apply the resources kind being listed?
What's special for `filter`?

Reply:
done